### PR TITLE
SE: Learn bool constraints from Equals with bool parameters.

### PR DESCRIPTION
**Pull request by mary-georgiou-sonarsource**
related to  #7704

### CODE REVIEW

File: analyzers/tests/SonarAnalyzer.UnitTest/TestCases/SymbolicExecution/Roslyn/ConditionEvaluatesToConstant.cs [hunk 1]
```diff
@@ -839,7 +839,7 @@ public void ReferenceEqualsBool()
         {
             bool a, b;
             a = b = true;
-            if (object.ReferenceEquals(a, b)) { }   // FN
+            if (object.ReferenceEquals(a, b)) { }   // Noncompliant
```
Comment (Tim-Pohlmann) [inline analyzers/tests/SonarAnalyzer.UnitTest/TestCases/SymbolicExecution/Roslyn/ConditionEvaluatesToConstant.cs:842]: Is this actually correct? Check the message to see whether this always evaluates to `true` or `false`.

**Review by Tim-Pohlmann — CHANGES_REQUESTED** [review]
Let's simplify this commit. Also add tests for the following:
- bool vs object
- bool vs null
- bool vs bool?
- bool? vs bool?
- bool? vs object
- bool? vs null

File: analyzers/tests/SonarAnalyzer.UnitTest/SymbolicExecution/Roslyn/RoslynSymbolicExecutionTest.Invocation.cs [hunk 2]
```diff
@@ -1199,6 +1199,65 @@ public void Main()
...
+        var validator = SETestContext.CreateCS(code).Validator;
+        validator.TagValue("Result").Should().HaveOnlyConstraints(ObjectConstraint.NotNull, BoolConstraint.From(expectedResult));
+        validator.TagValue("Left").Should().HaveOnlyConstraints(ObjectConstraint.NotNull, BoolConstraint.From(expectedConstraintLeft));
+        validator.TagValue("Right").Should().HaveOnlyConstraints(ObjectConstraint.NotNull, BoolConstraint.From(expectedConstraintRight));
+    }
+
+    [DataTestMethod]
+    [DataRow("true")]
+    [DataRow("false")]
+    public void Invocation_BoolEquals_Branches(string leftArgument)
```
[the remark this answers is not in the document]
Comment (mary-georgiou-sonarsource) [reply analyzers/tests/SonarAnalyzer.UnitTest/SymbolicExecution/Roslyn/RoslynSymbolicExecutionTest.Invocation.cs]: I just shownow that it learns nothing with only one constraint known.

File: analyzers/tests/SonarAnalyzer.UnitTest/SymbolicExecution/Roslyn/RoslynSymbolicExecutionTest.Invocation.cs [hunk 3]
```diff
@@ -1199,6 +1199,65 @@ public void Main()
...
+            },
+            x =>
+            {
+                x[result].HasConstraint(BoolConstraint.False).Should().BeTrue();
+                x[right].HasConstraint(x[left].Constraint<BoolConstraint>().Opposite).Should().BeTrue();
+            });
+    }
+
+    [DataTestMethod]
+    public void Invocation_BoolEquals_DoesNotLearn()
```
[the remark this answers is not in the document]
Comment (mary-georgiou-sonarsource) [reply analyzers/tests/SonarAnalyzer.UnitTest/SymbolicExecution/Roslyn/RoslynSymbolicExecutionTest.Invocation.cs]: I prefer to keep them separate these tests:
- I keep the test pattern as is for the other `EqualsTests`
- It's clearer the intend in every case - I show fast when it learns something and when it does not.

File: analyzers/src/SonarAnalyzer.Common/SymbolicExecution/Roslyn/OperationProcessors/Invocation.cs [hunk 4]
```diff
@@ -190,7 +190,13 @@ invocation switch
             _ => context.State.ToArray()
         };
 
-    private static ProgramState[] ProcessEquals(SymbolicContext context, IOperation leftOperation, IOperation rightOperation)
+    private static ProgramState[] ProcessEquals(SymbolicContext context, IOperation leftOperation, IOperation rightOperation) =>
+        context.State.Constraint<BoolConstraint>(leftOperation) is { }
+        && context.State.Constraint<BoolConstraint>(rightOperation) is { }
+            ? ProcessEqualsBool(context, leftOperation, rightOperation)
+            : ProcessEqualsObject(context, leftOperation, rightOperation);
```
Comment (Tim-Pohlmann) [inline analyzers/src/SonarAnalyzer.Common/SymbolicExecution/Roslyn/OperationProcessors/Invocation.cs:197]: Let's just inline this.
```suggestion
    private static ProgramState[] ProcessEquals(SymbolicContext context, IOperation leftOperation, IOperation rightOperation) =>
        context.State.Constraint<BoolConstraint>(leftOperation) is { } leftBool
        && context.State.Constraint<BoolConstraint>(rightOperation) is { } rightBool
            ? context.SetOperationConstraint(BoolConstraint.From(leftConstraint == rightConstraint)).ToArray()
            : ProcessEqualsObject(context, leftOperation, rightOperation);
```

File: analyzers/tests/SonarAnalyzer.UnitTest/SymbolicExecution/Roslyn/RoslynSymbolicExecutionTest.Invocation.cs [hunk 5]
```diff
@@ -1169,7 +1186,6 @@ public void Invocation_ReferenceEquals_SplitsToBothResults(string left, string r
                 x[arg].HasConstraint(ObjectConstraint.NotNull).Should().BeTrue();
             });
     }
-
```
Comment (Tim-Pohlmann) [inline analyzers/tests/SonarAnalyzer.UnitTest/SymbolicExecution/Roslyn/RoslynSymbolicExecutionTest.Invocation.cs:1172]: I assume you removed this empty line by mistake.

File: analyzers/tests/SonarAnalyzer.UnitTest/SymbolicExecution/Roslyn/RoslynSymbolicExecutionTest.Invocation.cs [hunk 6]
```diff
@@ -967,6 +967,21 @@ public void Invocation_ObjectEquals_LearnResult(string left, string right, bool
         validator.TagValue("Right").AllConstraints.Should().ContainSingle().Which.Kind.Should().Be(expectedConstraintRight);
     }
 
+    [DataTestMethod]
+    [DataRow("true", "null", false)]
+    [DataRow("null", "true", false)]
+    [DataRow("null", "(bool?)true", false)]
+    public void Invocation_ObjectEqualsMixedNullObjectArguments_LearnResult(string left, string right, bool expectedResult)
```
Comment (Tim-Pohlmann) [inline analyzers/tests/SonarAnalyzer.UnitTest/SymbolicExecution/Roslyn/RoslynSymbolicExecutionTest.Invocation.cs:974]: I think you can just incorporate this into `Invocation_BoolEquals_LearnsResult`.

File: analyzers/tests/SonarAnalyzer.UnitTest/TestCases/SymbolicExecution/Roslyn/ConditionEvaluatesToConstant.cs [hunk 7]
```diff
@@ -839,7 +839,7 @@ public void ReferenceEqualsBool()
         {
             bool a, b;
             a = b = true;
-            if (object.ReferenceEquals(a, b)) { }   // FN
+            if (object.ReferenceEquals(a, b)) { }   // Noncompliant {{Change this condition so that it does not always evaluate to 'True'.}}
```
Comment (Tim-Pohlmann) [inline analyzers/tests/SonarAnalyzer.UnitTest/TestCases/SymbolicExecution/Roslyn/ConditionEvaluatesToConstant.cs:842]: This always evaluates to false. The problem. You will need to adjust the `ReferenceEquals` to fix this.

**Review by Tim-Pohlmann — CHANGES_REQUESTED** [review]
Round 2

File: analyzers/tests/SonarAnalyzer.UnitTest/SymbolicExecution/Roslyn/RoslynSymbolicExecutionTest.Invocation.cs (hunk 6 above)
Comment (mary-georgiou-sonarsource) [reply analyzers/tests/SonarAnalyzer.UnitTest/SymbolicExecution/Roslyn/RoslynSymbolicExecutionTest.Invocation.cs:974]: It calls ObjectEquals and tests this method's behavior :/

File: analyzers/tests/SonarAnalyzer.UnitTest/TestCases/SymbolicExecution/Roslyn/ConditionEvaluatesToConstant.cs (hunk 7 above)
Comment (mary-georgiou-sonarsource) [reply analyzers/tests/SonarAnalyzer.UnitTest/TestCases/SymbolicExecution/Roslyn/ConditionEvaluatesToConstant.cs:842]: 🤦 Indeed. This should not even raise (The pr only deals with `Equals`(.

File: analyzers/tests/SonarAnalyzer.UnitTest/SymbolicExecution/Roslyn/RoslynSymbolicExecutionTest.Invocation.cs (hunk 6 above)
Comment (Tim-Pohlmann) [reply analyzers/tests/SonarAnalyzer.UnitTest/SymbolicExecution/Roslyn/RoslynSymbolicExecutionTest.Invocation.cs:974]: Maybe just add it to `Invocation_ObjectEquals_LearnResult` then.

Comment (mary-georgiou-sonarsource) [reply analyzers/tests/SonarAnalyzer.UnitTest/SymbolicExecution/Roslyn/RoslynSymbolicExecutionTest.Invocation.cs:974]: It's not straightforward - as discussed will leave as is.

**Pull request merged**